---
title: "ClusterPipeNov2019"
author: "Marisa Lim"
date: "11/4/2019"
output: html_document
---

Comparing results between tissue type and extraction methods for:

- ONT barcode assignment per sequence run per flowcell:
  
  - qcat demultiplexing

  - minibar demultiplexing

- species ID stats 100 vs. 500 vs. 5,000 subsets

```{r, echo=FALSE, message=FALSE}
library(reshape2)
library(plyr)
library(ggplot2)
library(cowplot)
library(scales)
library(seqinr)
library(dplyr)
library(kableExtra)
library(stringr)

setwd('~/Desktop/SNOWDUCK_ANALYSES/pipereanalysisNov2019/')
```

# Figure 2: MiniBar vs. qcat demultiplexed read barcode assignments
```{r, echo=FALSE}
# Qcat ONT assignment input file: 
# qcat results
qcat_dat <- read.csv('fulldatqcat99.csv', header=TRUE, stringsAsFactors = FALSE)

qcat_dat$grid_factor <- factor(qcat_dat$Flowcell, levels=c('FAL19910', 'FAL19272'))
qcat_dat$demult <- 'qcat'
qcat_dat$Barcode[qcat_dat$Barcode=='bc01'] <- 'barcode01'
qcat_dat$Barcode[qcat_dat$Barcode=='bc02'] <- 'barcode02'
qcat_dat$Barcode[qcat_dat$Barcode=='bc03'] <- 'barcode03'
qcat_dat$Barcode[qcat_dat$Barcode=='bc04'] <- 'barcode04'
qcat_dat$Barcode[qcat_dat$Barcode=='bc05'] <- 'barcode05'
qcat_dat$Barcode[qcat_dat$Barcode=='bc06'] <- 'barcode06'
qcat_dat$Barcode[qcat_dat$Barcode=='bc07'] <- 'barcode07'
qcat_dat$Barcode[qcat_dat$Barcode=='bc08'] <- 'barcode08'
qcat_dat$Barcode[qcat_dat$Barcode=='bc09'] <- 'barcode09'
qcat_dat$Barcode[qcat_dat$Barcode=='bc10'] <- 'barcode10'
qcat_dat$Barcode[qcat_dat$Barcode=='bc11'] <- 'barcode11'
qcat_dat$Barcode[qcat_dat$Barcode=='bc12'] <- 'barcode12'

# kable(qcat_dat, align='l') %>%
#   kable_styling(bootstrap_options = c('striped', 'hover')) %>%
#   scroll_box(width='100%', height='300px')
```

```{r, echo=FALSE}
# MiniBar ONT assignment input file: 
# minibar results
# these files were generated by running: bash minibar_batcher.sh
mb_0906 <- read.table('minibar_ONTbarcodeassignment/20190906_minibarcounts.txt', header=TRUE)
mb_0909 <- read.table('minibar_ONTbarcodeassignment/20190909_minibarcounts.txt', header=TRUE)
mb_0911 <- read.table('minibar_ONTbarcodeassignment/20190911_minibarcounts.txt', header=TRUE)
mb_0913 <- read.table('minibar_ONTbarcodeassignment/20190913_minibarcounts.txt', header=TRUE)

# sum across barcodes since minibar splits files
# add in 0s for absent barcodes
barcodes <- c('barcode01', 'barcode02', 'barcode03', 'barcode04', 'barcode05', 
              'barcode06', 'barcode07', 'barcode08', 'barcode09', 'barcode10',
              'barcode11', 'barcode12', 'unk')

minibar_parsing <- function(mbdat){
  mb_fmt <- data.frame('Num_reads'=0, 'Barcode'='thebarcode', 'Dataset'='dat', 
                         'Flowcell'='flowcell', 'Seq_run'='seqrun', 'Per_reads'=0)
  for(i in 1:length(barcodes)){
    thebarcode <- barcodes[i]
    thedat <- mbdat[mbdat$Barcode == thebarcode, ]
    if(nrow(thedat) == 0){
      newrow <- data.frame('Num_reads'=0, 'Barcode'=thebarcode, 'Dataset'=as.character(mbdat$Dataset[1]), 
                         'Flowcell'=mbdat$Flowcell[1], 'Seq_run'=mbdat$Seq_run[1], 'Per_reads'=0)
      mb_fmt <- rbind(mb_fmt, newrow)
    }else if(nrow(thedat) != 0){
      summed_reads <- ddply(thedat, 'Barcode', numcolwise(sum))
      newrow <- data.frame('Num_reads'=summed_reads$Num_reads, 'Barcode'=thebarcode, 'Dataset'=as.character(thedat$Dataset[1]), 
                         'Flowcell'=thedat$Flowcell[1], 'Seq_run'=thedat$Seq_run[1], 'Per_reads'=summed_reads$Per_reads)
      mb_fmt <- rbind(mb_fmt, newrow)
    }
  }
  thembdat <- mb_fmt[-1,]
  return(thembdat)
}

mb_0906_fmt <- minibar_parsing(mb_0906)
mb_0909_fmt <- minibar_parsing(mb_0909)
mb_0911_fmt <- minibar_parsing(mb_0911)
mb_0913_fmt <- minibar_parsing(mb_0913)

all_mb_dat <- rbind(mb_0906_fmt, mb_0909_fmt, mb_0911_fmt, mb_0913_fmt)
all_mb_dat$grid_factor <- factor(all_mb_dat$Flowcell, levels=c('FAL19910', 'FAL19272'))
all_mb_dat2 <- all_mb_dat[, c(3:5, 2, 1, 6)]
all_mb_dat2$demult <- 'minibar'

# kable(all_mb_dat2, align='l') %>%
#   kable_styling(bootstrap_options = c('striped', 'hover')) %>%
#   scroll_box(width='100%', height='300px')
```

```{r, echo=FALSE, fig.width=10}
all_demult <- rbind(all_mb_dat2, qcat_dat[,c(1:6,9)])

all_demult2 <- all_demult[all_demult$Barcode != 'none' & all_demult$Barcode != 'unk', ]
print('Input data:')
kable(all_demult2, align='l') %>%
  kable_styling(bootstrap_options = c('striped', 'hover')) %>%
  scroll_box(width='100%', height='300px')

ggplot(all_demult2, aes(x=Barcode, y=Num_reads, shape=Seq_run)) +
  facet_grid(Flowcell~demult, scales='free') + geom_point(size=4, alpha=0.7) +
  geom_line(aes(group=Dataset), alpha=0.4, lty=2) +
  scale_y_continuous(labels=scales::comma) +
  ylab('# Assigned reads') + xlab('ONT Barcode') + labs(shape='Sequence run', color='Dataset') +
  theme(axis.text.x=element_text(angle=45, hjust=1), panel.grid.major=element_line(size=0.5, colour='seashell'), legend.position='top')
ggsave('./snowduck_figures/demult_qcatVSminibar_numreads.jpg', height=6, width=8, units='in', dpi=500)

ggplot(all_demult2, aes(x=Barcode, y=Per_reads/100, shape=Seq_run)) +
  facet_grid(Flowcell~demult, scales='free') + geom_point(size=4, alpha=0.7) + 
  geom_line(aes(group=Dataset), alpha=0.4, lty=2) +
  scale_y_continuous(labels=scales::percent) +
  ylab('% Assigned reads') + xlab('ONT Barcode') + labs(shape='Sequence run', color='Dataset') +
  theme(axis.text.x=element_text(angle=45, hjust=1), panel.grid.major=element_line(size=0.5, colour='seashell'), legend.position='top')
ggsave('./snowduck_figures/demult_qcatVSminibar_percentreads.jpg', height=6, width=8, units='in', dpi=500) 
```

# Figures 3-4: **Minibar vs. qcat** subset results

```{r, echo=FALSE}
# load input files for subset sp IDs
spID_parser <- function(filepath, datID, tissuetype, demult){
  thedat <- read.table(filepath, header=TRUE, sep='\t')
  thedat$dat <- datID
  thedat$demult <- demult
  thedat$subset <- paste(substring(str_split(filepath, '_', simplify=TRUE)[,3], nchar(demult)+1), ' reads', sep='')
  thedat$tissue_type <- tissuetype
  thedat$extr_method <- substr(str_split(thedat$Sample, tissuetype, simplify = TRUE)[,2], 1, 2)
  thedat$trueID <- substring(str_split(thedat$Sample, tissuetype, simplify = TRUE)[,2], 3)
  df <- data.frame(thedat[1,], 'SampleID'='SampleID')
  for(i in 1:nrow(thedat)){
    therow <- thedat[i,]
    therow$SampleID <- paste(therow$tissue_type, substr(str_split(therow$Sample, therow$tissue_type, simplify=TRUE)[,2], 1, 1), sep='_')
    df <- rbind(df, therow)
  }
  df2 <- df[-1,]
  return(df2)
}

# minibar
dat0906_100mb <- spID_parser('~/Desktop/clstrfq2019/4_spID/20190906_minibar100_allsamps_parsedout.txt', '20190906', 'FFPE', 'minibar')
dat0906_500mb <- spID_parser('~/Desktop/clstrfq2019/4_spID/20190906_minibar500_allsamps_parsedout.txt', '20190906', 'FFPE', 'minibar')
dat0906_5kmb <- spID_parser('~/Desktop/clstrfq2019/4_spID/20190906_minibar5000_allsamps_parsedout.txt', '20190906', 'FFPE', 'minibar')

dat0911_100mb <- spID_parser('~/Desktop/clstrfq2019/4_spID/20190911_minibar100_allsamps_parsedout.txt', '20190911', 'poo', 'minibar')
dat0911_500mb <- spID_parser('~/Desktop/clstrfq2019/4_spID/20190911_minibar500_allsamps_parsedout.txt', '20190911', 'poo', 'minibar')
dat0911_5kmb <- spID_parser('~/Desktop/clstrfq2019/4_spID/20190911_minibar5000_allsamps_parsedout.txt', '20190911', 'poo', 'minibar')

dat0913_100mb <- spID_parser('~/Desktop/clstrfq2019/4_spID/20190913_minibar100_allsamps_parsedout.txt', '20190913', 'liver', 'minibar')
dat0913_500mb <- spID_parser('~/Desktop/clstrfq2019/4_spID/20190913_minibar500_allsamps_parsedout.txt', '20190913', 'liver', 'minibar')
dat0913_5kmb <- spID_parser('~/Desktop/clstrfq2019/4_spID/20190913_minibar5000_allsamps_parsedout.txt', '20190913', 'liver', 'minibar')

# qcat
dat0906_100qc <- spID_parser('~/Desktop/clstrfq2019/4_spID/20190906_qcat100_allsamps_parsedout.txt', '20190906', 'FFPE', 'qcat')
dat0906_500qc <- spID_parser('~/Desktop/clstrfq2019/4_spID/20190906_qcat500_allsamps_parsedout.txt', '20190906', 'FFPE', 'qcat')
dat0906_5kqc <- spID_parser('~/Desktop/clstrfq2019/4_spID/20190906_qcat5000_allsamps_parsedout.txt', '20190906', 'FFPE', 'qcat')

dat0911_100qc <- spID_parser('~/Desktop/clstrfq2019/4_spID/20190911_qcat100_allsamps_parsedout.txt', '20190911', 'poo', 'qcat')
dat0911_500qc <- spID_parser('~/Desktop/clstrfq2019/4_spID/20190911_qcat500_allsamps_parsedout.txt', '20190911', 'poo', 'qcat')
dat0911_5kqc <- spID_parser('~/Desktop/clstrfq2019/4_spID/20190911_qcat5000_allsamps_parsedout.txt', '20190911', 'poo', 'qcat')

dat0913_100qc <- spID_parser('~/Desktop/clstrfq2019/4_spID/20190913_qcat100_allsamps_parsedout.txt', '20190913', 'liver', 'qcat')
dat0913_500qc <- spID_parser('~/Desktop/clstrfq2019/4_spID/20190913_qcat500_allsamps_parsedout.txt', '20190913', 'liver', 'qcat')
dat0913_5kqc <- spID_parser('~/Desktop/clstrfq2019/4_spID/20190913_qcat5000_allsamps_parsedout.txt', '20190913', 'liver', 'qcat')

# 0909 dat has feather and hair tissue types, so can't imput simply as a string
spID_parse_modified <- function(filepath, datID, demult){
  thedat <- read.table(filepath, header=TRUE, sep='\t')
  thedat$dat <- datID
  thedat$demult <- demult
  thedat$subset <- paste(substring(str_split(filepath, '_', simplify=TRUE)[,3], nchar(demult)+1), ' reads', sep='')
  thedat$tissue_type <- c(rep('feather', 3), rep('hair', 3)) # assumes that 1st three are feather, last 3 are hair
  df <- data.frame(thedat[1,],'extr_method'='extr_method', 'trueID'='trueID', 'SampleID'='SampleID')
  for(i in 1:nrow(thedat)){
    therow <- thedat[i,]
    therow$extr_method <- substr(str_split(therow$Sample, therow$tissue_type, simplify=TRUE)[,2], 1, 2)
    therow$trueID <- substring(str_split(therow$Sample, therow$tissue_type, simplify=TRUE)[,2], 3)
    therow$SampleID <- paste(therow$tissue_type, substr(str_split(therow$Sample, therow$tissue_type, simplify=TRUE)[,2], 1, 1), sep='_')
    df <- rbind(df, therow)
  }
  df2 <- df[-1,]
  return(df2)
}

dat0909_100mb <- spID_parse_modified('~/Desktop/clstrfq2019/4_spID/20190909_minibar100_allsamps_parsedout.txt', '20190909', 'minibar')
dat0909_500mb <- spID_parse_modified('~/Desktop/clstrfq2019/4_spID/20190909_minibar500_allsamps_parsedout.txt', '20190909', 'minibar')
dat0909_5kmb <- spID_parse_modified('~/Desktop/clstrfq2019/4_spID/20190909_minibar5000_allsamps_parsedout.txt', '20190909', 'minibar')

dat0909_100qc <- spID_parse_modified('~/Desktop/clstrfq2019/4_spID/20190909_qcat100_allsamps_parsedout.txt', '20190909', 'qcat')
dat0909_500qc <- spID_parse_modified('~/Desktop/clstrfq2019/4_spID/20190909_qcat500_allsamps_parsedout.txt', '20190909', 'qcat')
dat0909_5kqc <- spID_parse_modified('~/Desktop/clstrfq2019/4_spID/20190909_qcat5000_allsamps_parsedout.txt', '20190909', 'qcat')
```

```{r, echo=FALSE}
# combine dfs
all_0906_subset <- rbind(dat0906_100mb, dat0906_500mb, dat0906_5kmb, dat0906_100qc, dat0906_500qc, dat0906_5kqc)
all_0909_subset <- rbind(dat0909_100mb, dat0909_500mb, dat0909_5kmb, dat0909_100qc, dat0909_500qc, dat0909_5kqc)
all_0911_subset <- rbind(dat0911_100mb, dat0911_500mb, dat0911_5kmb, dat0911_100qc, dat0911_500qc, dat0911_5kqc)
all_0913_subset <- rbind(dat0913_100mb, dat0913_500mb, dat0913_5kmb, dat0913_100qc, dat0913_500qc, dat0913_5kqc)

all_dat_subs <- rbind(all_0906_subset, all_0909_subset, all_0911_subset, all_0913_subset)
all_dat_subs$trueID <- gsub('leo', 'snow_leopard', all_dat_subs$trueID)
all_dat_subs$trueID <- gsub('duck', 'cinnamon_teal', all_dat_subs$trueID)
all_dat_subs$tissue_type <- gsub('poo', 'scat', all_dat_subs$tissue_type)
all_dat_subs$SampleID <- gsub('poo_', 'scat_', all_dat_subs$SampleID)

all_dat_subs$extr_method <- gsub('qi', 'qiagen', all_dat_subs$extr_method)
all_dat_subs$extr_method <- gsub('bi', 'biomeme', all_dat_subs$extr_method)
all_dat_subs$extr_method <- gsub('ch', 'chelex', all_dat_subs$extr_method)

kable(all_dat_subs, align='l') %>%
  kable_styling(bootstrap_options = c('striped', 'hover')) %>%
  scroll_box(width='100%', height='300px')
```

```{r, echo=FALSE, fig.width=10}
fillcolors <- c("red", "blue")

ggplot(all_dat_subs, aes(x=SampleID, y=per_id/100, shape=extr_method, col=demult, group=SampleID)) +
  facet_grid(subset~trueID, scales='free_x') +
  geom_point(size=4, alpha=0.6) + 
  geom_line(alpha=0.2, col='black') +
  geom_hline(yintercept=0.99, lty=2, alpha=0.6, col='red') +
  geom_vline(data=filter(all_dat_subs, trueID=='snow_leopard'), aes(xintercept=4), lty=2, alpha=0.6) +
  geom_vline(data=filter(all_dat_subs, trueID=='snow_leopard'), aes(xintercept=6), lty=2, alpha=0.6) +
  geom_vline(data=filter(all_dat_subs, trueID=='snow_leopard'), aes(xintercept=9), lty=2, alpha=0.6) +
  scale_color_manual(values=fillcolors) +
  scale_y_continuous(limits=c(0.985,1),
                     labels=scales::percent) +
  ylab('% Sequence ID') + xlab('Sample') +
  labs(shape='Extraction method', col='Demultiplexer') +
  theme(axis.text.x=element_text(angle=45, hjust=1), panel.grid.major=element_line(size=0.5, colour='seashell'), legend.position='top')
ggsave('./snowduck_figures/spID_qcatVSminibar_100VS500VS5kreads.jpg', height=6, width=10, units='in', dpi=500)

ggplot(all_dat_subs, aes(x=SampleID, y=prop_demreads_inclstr, shape=extr_method, col=demult, group=SampleID)) +
  facet_grid(subset~trueID, scales='free_x') +
  geom_point(size=4, alpha=0.6) + 
  geom_line(alpha=0.2, col='black') +
  geom_vline(data=filter(all_dat_subs, trueID=='snow_leopard'), aes(xintercept=4), lty=2, alpha=0.6) +
  geom_vline(data=filter(all_dat_subs, trueID=='snow_leopard'), aes(xintercept=6), lty=2, alpha=0.6) +
  geom_vline(data=filter(all_dat_subs, trueID=='snow_leopard'), aes(xintercept=9), lty=2, alpha=0.6) +
  scale_color_manual(values=fillcolors) +
  scale_y_continuous(limits=c(0.1,1), 
                     labels=scales::percent) +
  ylab('% Demultiplexed reads clustered') + xlab('Sample') +
  labs(shape='Extraction method', col='Demultiplexer') +
  theme(axis.text.x=element_text(angle=45, hjust=1), panel.grid.major=element_line(size=0.5, colour='seashell'), legend.position='top')
ggsave('./snowduck_figures/propclstr_qcatVSminibar_100VS500VS5kreads.jpg', height=6, width=10, units='in', dpi=500)

ggplot(all_dat_subs, aes(x=SampleID, y=aln_len, shape=extr_method, col=demult, group=SampleID)) +
  facet_grid(subset~trueID, scales='free_x') +
  geom_point(size=4, alpha=0.6) + 
  geom_line(alpha=0.2, col='black') +
  geom_vline(data=filter(all_dat_subs, trueID=='snow_leopard'), aes(xintercept=4), lty=2, alpha=0.6) +
  geom_vline(data=filter(all_dat_subs, trueID=='snow_leopard'), aes(xintercept=6), lty=2, alpha=0.6) +
  geom_vline(data=filter(all_dat_subs, trueID=='snow_leopard'), aes(xintercept=9), lty=2, alpha=0.6) +
  scale_color_manual(values=fillcolors) +
  ylab('Alignment length (bp)') + xlab('Sample') +
  labs(shape='Extraction method', col='Demultiplexer') +
  theme(axis.text.x=element_text(angle=45, hjust=1), panel.grid.major=element_line(size=0.5, colour='seashell'), legend.position='top')
ggsave('./snowduck_figures/alnlen_qcatVSminibar_100VS500VS5kreads.jpg', height=6, width=10, units='in', dpi=500)
```

# For Figure 5: Extract consensus sequences to make Mafft alignment (in Geneious)

```{r, echo=FALSE}
mycons <- all_dat_subs[, c(1, 19:26)]
mycons$fastaheader <- paste(mycons$trueID, mycons$tissue_type, mycons$extr_method, mycons$demult, mycons$subset, sep="_")
mycons_duck <- mycons[mycons$trueID == 'cinnamon_teal',]
mycons_leo <- mycons[mycons$trueID == 'snow_leopard',]

kable(mycons_duck, align='l') %>%
  kable_styling(bootstrap_options = c('striped', 'hover')) %>%
  scroll_box(width='100%', height='300px')

kable(mycons_leo, align='l') %>%
  kable_styling(bootstrap_options = c('striped', 'hover')) %>%
  scroll_box(width='100%', height='300px')

write.fasta(sequences=as.list(mycons_duck$MkConsSeq), names=mycons_duck$fastaheader, 'duck_allsubsetconsseqs.fasta', open='w', as.string=FALSE)
write.fasta(sequences=as.list(mycons_leo$MkConsSeq), names=mycons_leo$fastaheader, 'leopard_allsubsetconsseqs.fasta', open='w', as.string=FALSE)
```

# Table 1: MiniBar vs. qcat demultiplexed read stats

```{r, echo=FALSE}
# full data demultiplexer nano stats
nanostatfiles <- list.files('~/Desktop/clstrfq2019/2b_demultiplexed/', pattern='_nanostats.txt')

nsdf <- data.frame('Mean read length'=1,	'Mean read quality'=1,	'Median read length'=1,	'Median read quality'=1,	'Number of reads'=1,	'Read length N50'=1,	'Total bases'=1,	'SampID'='id',	'Demult'='demult')
for(i in 1:length(nanostatfiles)){
  thedat <- read.table(paste('~/Desktop/clstrfq2019/2b_demultiplexed/', nanostatfiles[i], sep=''), header=TRUE, sep='\t')
  nsdf <- rbind(nsdf, thedat)
}
nsdf2 <- nsdf[-1, c(8,9,1:2,5,3:4,6:7)]

kable(nsdf2, align='l') %>%
  kable_styling(bootstrap_options = c('striped', 'hover')) %>%
  scroll_box(width='100%', height='300px')

write.csv(bydemult_mean, './snowduck_figures/mbVSqcat_demultnanostats.csv', row.names=FALSE)
```

# Table 2: Calculate average and standard deviation stats per demultiplexer/subset

```{r, echo=FALSE}
#by demult per subset
all_dat_subs_counts1 <- melt(all_dat_subs[ , c(4, 5, 8, 9, 20:22)], id.var=c('dat', 'subset', 'demult'))
subset_avgcounts_across_all_dats1 <- ddply(all_dat_subs_counts1, c('subset', 'demult', 'variable'), summarise, mean=mean(value), sd=sd(value))
bydemult_mean <- dcast(subset_avgcounts_across_all_dats1, subset + demult ~ variable, value.var='mean')
bydemult_sd <- dcast(subset_avgcounts_across_all_dats1, subset + demult ~ variable, value.var='sd')
kable(bydemult_mean, align='l') %>%
  kable_styling(bootstrap_options = c('striped', 'hover')) 
kable(bydemult_sd, align='l') %>%
  kable_styling(bootstrap_options = c('striped', 'hover')) 

write.csv(bydemult_mean, './snowduck_figures/AVGspIDmetsBYdemult_alldats.csv', row.names=FALSE)
write.csv(bydemult_sd, './snowduck_figures/SDspIDmetsBYdemult_alldats.csv', row.names=FALSE)
```

# Table S4: Calculate average and standard deviation stats per extraction method/demultiplexer/subset

```{r, echo=FALSE}
#by extr method per demult per subset
all_dat_subs_counts2 <- melt(all_dat_subs[ , c(4, 5, 8, 9, 20:22, 24)], id.var=c('dat', 'subset', 'demult', 'extr_method'))
subset_avgcounts_across_all_dats2 <- ddply(all_dat_subs_counts2, c('subset', 'demult', 'extr_method', 'variable'), summarise, mean=mean(value), sd=sd(value))
byextr_mean <- dcast(subset_avgcounts_across_all_dats2, subset + demult + extr_method ~ variable, value.var='mean')
byextr_sd <- dcast(subset_avgcounts_across_all_dats2, subset + demult + extr_method ~ variable, value.var='sd')
kable(byextr_mean, align='l') %>%
  kable_styling(bootstrap_options = c('striped', 'hover')) 
kable(byextr_sd, align='l') %>%
  kable_styling(bootstrap_options = c('striped', 'hover')) 

write.csv(byextr_mean, './snowduck_figures/AVGspIDmetsBYextrmeth_alldats.csv', row.names=FALSE)
write.csv(byextr_sd, './snowduck_figures/SDspIDmetsBYextrmeth_alldats.csv', row.names=FALSE)
```
